### ISSUE TITLE
Add Rice Distribution 

### ISSUE THREAD

**Issue by AndreasMittermeier** [pymc-devs/pymc#1638]
I use pymc3 to evaluate MRI IVIM data, which has a rice distributed signal. As the rice distribution is missing in  pymc3/distributions/continuous.py, I tried implementing it myself. Here is the file I located at my local pymc3/distributions/ folder:

```
from pymc3.distributions.distribution import Continuous, draw_values, generate_samples
from pymc3.distributions.dist_math import bound, logpow, gammaln, betaln, std_cdf, i0, i1
import numpy as np
import theano.tensor as tt
from scipy import stats

class Rice(Continuous):
    R"""
    Rice distribution.

    .. math::

       f(x\mid \nu ,\sigma )=
       {\frac  {x}{\sigma ^{2}}}\exp
       \left({\frac  {-(x^{2}+\nu ^{2})}{2\sigma ^{2}}}\right)I_{0}\left({\frac  {x\nu }{\sigma ^{2}}}\right),

    ========  ==============================================================
    Support   :math:`x \in (0, +\infinity)`
    Mean      :math:`\sigma {\sqrt  {\pi /2}}\,\,L_{{1/2}}(-\nu ^{2}/2\sigma ^{2})`
    Variance  :math:`2\sigma ^{2}+\nu ^{2}-{\frac  {\pi \sigma ^{2}}{2}}L_{{1/2}}^{2}
                        \left({\frac  {-\nu ^{2}}{2\sigma ^{2}}}\right)`
    ========  ==============================================================


    Parameters
    ----------
    nu : float
        shape parameter.
    sd : float
        standard deviation.

    """

    def __init__(self, sd=None, nu = None, *args, **kwargs):
        super(Rice, self).__init__(*args, **kwargs)
        self.sd = sd
        self.nu = nu


    def random(self, point=None, size=None, repeat=None):
        sd, nu = draw_values([self.sd, self.nu],
                                  point=point)
        return generate_samples(stats.rice.rvs, sd, nu,
                                dist_shape=self.shape,
                                size=size)

    def logp(self, value):
        sd = self.sd
        nu = self.nu
        return tt.log(value/sd**2*tt.exp(-(value**2+nu**2)/(2*sd**2))*i0(value*nu/sd**2))
```

In order to validate my implementation, I simulated signal intensity curves that are rice distributed and follow a biexponential decay. I assume these curves belong to different pixels in a region of interest and that all pixels share a common set of parameters (f, D ,Dstar), which define the shape of the curve. To generate these curves, I draw samples from a predefined distribution and put these parameters into a signal model :

```
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pymc3 as pm
from pymc3.distributions.RiceDist import Rice
import csv
import matplotlib
% matplotlib notebook

# draw values from priors
np.random.seed(51)
size=10
fig=plt.figure()
with pm.Model() as model:
    f = pm.Beta('f',mu=0.15,sd=0.10).random(size=size)
    D = pm.Gamma('D', mu=1.8, sd=0.9).random(size=size)
    Dstar = pm.Gamma('Dstar', mu=10, sd=5).random(size=size)
    plt.hist(f, bins=100, range=(0,1), alpha=1)
    #plt.hist(D, bins=100, range=(0,4), alpha=1)
    #plt.hist(Dstar, bins=100, range=(0,20), alpha=1)
# simulate b values and signal
b = pd.DataFrame([[0,10,30,50,80,120,200,350,500,800]]*size).T
S = (1-f)*np.exp(-b*D/1000)+f*np.exp(-b*(D+Dstar)/1000)

# make signal rice distributed
sig = 0.01
def ricernd(data, noise):
    x = noise * np.random.randn(len(data.columns)) + data
    y = noise * np.random.randn(len(data.columns))
    r = np.sqrt(x ** 2 + y ** 2)
    return r
S = ricernd(S, sig)
```

The signal intensity curves plotted:

![index](https://cloud.githubusercontent.com/assets/9657209/21590398/c587180e-d0fb-11e6-9470-bc59205d611a.png)

Then I tried to evaluate the data like in @twiecki 's blog about hierarchical modeling (http://twiecki.github.io/blog/2014/03/17/bayesian-glms-3/) with a pooled, unpooled and partially pooled model. 
For the pooled model everything works out fine, meaning the sampler converges and the posterior for each parameter looks like the distribution I draw the samples from:

```
# initialize Model
with pm.Model() as biexp_pooled_model:
    # Priors
    f = pm.Beta('f', mu=0.15,sd=0.10)
    D = pm.Gamma('D', mu=1.8,sd=0.9)
    Dstar = pm.Gamma('Dstar',mu=10,sd=5)
    sig = pm.HalfNormal('sigma')

    # Model prediction
    exp1=-(b.values*D/1000)
    exp2=-(b.values*(D+Dstar)/1000)
    E_S = (1-f)*np.exp(exp1)+f*np.exp(exp2)
    
    # Likelihood of observations
    S_obs = Rice('S_patient_obs', nu=E_S, sd=sig, observed=S.values)  
```
```
# sample and save trace
with biexp_pooled_model:
    start = pm.find_MAP()
    print start
    pooled_trace = pm.sample(2000,start=start, njobs=8)
    pm.backends.text.dump('/home/amitter/IVIMBayes/Sim_Data_Biexp/trace_pooled_rice/', pooled_trace)
```
The traceplot looks like this:

![index1](https://cloud.githubusercontent.com/assets/9657209/21590466/52f44554-d0fc-11e6-8132-07b982add1a6.png)

But when i analyze each signal curve by itself, as in the unpooled model, the pm.sample doesn't converge. As an example I took the first signal curve S[0]:

```
# initialize Model
with pm.Model() as biexp_pooled_model:
    # Priors
    f = pm.Beta('f', mu=0.15,sd=0.10)
    D = pm.Gamma('D', mu=1.8,sd=0.9)
    Dstar = pm.Gamma('Dstar',mu=10,sd=5)
    sig = pm.HalfNormal('sigma')

    # Model prediction
    exp1=-(b[0].values*D/1000)
    exp2=-(b[0].values*(D+Dstar)/1000)
    E_S = (1-f)*np.exp(exp1)+f*np.exp(exp2)
    
    # Likelihood of observations
    S_obs = Rice('S_patient_obs', nu=E_S, sd=sig, observed=S[0].values)  
```
Sampling is the same as above, traceplot looks like this:

![index2](https://cloud.githubusercontent.com/assets/9657209/21590494/8f07a8d8-d0fc-11e6-89e7-aabbc90281ae.png)

 Also for the partially pooled model, the sampler does not converge. Any idea why is could be?
(When i use pm.Normal or pm.StudentT instead of Rice for the likelihood of observations, the sampler converges, but this does not reflect the correct nature of the signal. Increasing the number of curves does not help either.)
Thanks in advance and sorry for the long post.

**Comment by fonnesbeck** [comment]
Try letting go of the MAP initialization and let PyMC initialize it using ADVI (which it will do by default). I often find MAP Does not provide a good start, as counterintuitive as that seems).

**Comment by vik975** [comment]
can i work on it ?

**Comment by souravsingh** [comment]
The issue can be closed now.

**Comment by ColCarroll** [comment]
#1819

**Comment by twiecki** [comment]
https://github.com/pymc-devs/pymc3/pull/1862

**Comment by MBrouns** [comment]
I made an attempt in #2963 but still encounter some issues, would anyone be willing to give some pointers?

**Comment by MBrouns** [comment]
I guess this one can be closed with #2964

**Comment by junpenglao** [comment]
Thanks @MBrouns!
